### PR TITLE
Fix #10064 - EPW file not working if it doesn't have an ending newline

### LINKED ISSUE #10064: EPW file not working if it doesn't have an ending newline
An otherwise perfectly valid weather file breaks the simulation if it doesn't end in a newline.

```
**  Fatal  ** Unexpected error condition in middle of reading EPW file, stopped at 2007/1/1 1:60
   ...Summary of Errors that led to program termination:
   ..... Reference severe error count=0
   ..... Last severe error=
```

This used to work fine in older E+ versions and the culprit is probably this line[1], if the readLine() invocation expects all lines to end in a newline, as it would miss the last one.

[1] https://github.com/NREL/EnergyPlus/blob/b03805c97d6511a3232300e65b0d30dc0d627816/src/EnergyPlus/WeatherManager.cc#L2682

I tested this using the OpenStudioApplication 1.6 (EnergyPlus 23.1.0) on an Ubuntu 22.04 system.

The attached file has two identical epw  files except for the ending newline to show the problem.

[D3_peninsula_bug_.zip](https://github.com/NREL/EnergyPlus/files/11690315/D3_peninsula_bug_.zip)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fixes #10064

https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/10000s/10064

Running 1ZoneUncontrolled with original defect EPW.
before it fails. 

```
   **  Fatal  ** Unexpected error condition in middle of reading EPW file, stopped at 2007/1/1 1:60
```
After it works.

I also ran with the EPW that had a newline using develop, compared to the after fix, same weather output variables.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/IOFiles.hh [hunk 1]
```diff
@@ -531,6 +531,9 @@ public:
     };
 
     void close();
+
+    // This is different from istream::good(), which is false if EOF is true while there were no errors (happens when no EOL at end of file)
+    // this operate like `operator bool(istream& is)` <=> `!is.bad() && !is.fail()`
```
Comment (jmarrec) [inline src/EnergyPlus/IOFiles.hh:536]: new docstring for `ReadResult::good()` explaining the change.

Refer to the table in "See Also" at https://en.cppreference.com/w/cpp/io/ios_base/iostate

image

File: src/EnergyPlus/IOFiles.cc [hunk 2]
```diff
@@ -89,15 +89,19 @@ void InputFile::close()
...
+    if (std::getline(*is, line)) {
         if (!line.empty() && line.back() == '\r') {
             line.pop_back();
         }
-        return {std::move(line), is->eof(), is->good()};
+        // Use operator bool, see ReadResult::good() docstring
+        return {std::move(line), is->eof(), bool(is)};
     } else {
-        return {"", true, false};
+        return {"", is->eof(), false};
```
Comment (jmarrec) [inline src/EnergyPlus/IOFiles.cc:104]: Actual fix is here.

File: src/EnergyPlus/IOFiles.hh [hunk 3]
```diff
@@ -563,7 +566,8 @@ public:
         if (is) {
             T result;
             *is >> result;
-            return ReadResult<T>{result, is->eof(), is->good()};
+            // Use operator bool, see ReadResult::good() docstring
+            return ReadResult<T>{result, is->eof(), bool(is)};
```
Comment (jmarrec) [inline src/EnergyPlus/IOFiles.hh:570]: Consistency.

File: tst/EnergyPlus/unit/WeatherManager.unit.cc [hunk 4]
```diff
@@ -2446,3 +2446,66 @@ TEST_F(EnergyPlusFixture, DisplayWeatherMissingDataWarnings_TMYx)
...
+    bool Available = false;
+    bool ErrorsFound = false;
+    WeatherManager::GetNextEnvironment(*state, Available, ErrorsFound); // Does not throw
+    ASSERT_FALSE(ErrorsFound);
+    EXPECT_TRUE(compare_err_stream("", true));
+    EXPECT_EQ(1, state->dataWeatherManager->NumOfEnvrn);
+    EXPECT_TRUE(compare_enums(state->dataWeatherManager->Environment(1).KindOfEnvrn, Constant::KindOfSim::RunPeriodWeather));
+
+    EXPECT_NO_THROW(ReadWeatherForDay(*state, 1, 1, true));
+    EXPECT_TRUE(compare_err_stream("", true));
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/WeatherManager.unit.cc:2510]: This throws beforehand, with the `Unexpected error condition in middle of reading EPW file` error

File: tst/EnergyPlus/unit/Resources/chicago_no_eol_at_end_of_file.epw
Comment (jmarrec) [inline tst/EnergyPlus/unit/Resources/chicago_no_eol_at_end_of_file.epw:1]: This is USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw, and did `truncate -s -1 chicago_no_eol_at_end_of_file.epw`

File: tst/EnergyPlus/unit/WeatherManager.unit.cc (hunk 4 above)
Comment (Myoldmopar) [reply tst/EnergyPlus/unit/WeatherManager.unit.cc:2510]: Good test, although I wonder if you really needed the input objects listed here.

File: src/EnergyPlus/IOFiles.hh (hunk 1 above)
Comment (Myoldmopar) [reply src/EnergyPlus/IOFiles.hh:536]: Thanks for the clarification.

**Review by Myoldmopar — APPROVED** [review]
Testing it out now, thanks @jmarrec

**Comment by Myoldmopar** [comment]
Everything works including the new unit test.  This is good to go, thanks @jmarrec

**Pull request merged**